evm/src/cpu/decode.rs

```diff
@@ -0,0 +1,281 @@
+    (0xc0, 5, columns::IS_INVALID_14), // 0xc0-0xdf
+    (0xe0, 4, columns::IS_INVALID_15), // 0xe0-0xef
+    (0xf0, 0, columns::IS_CREATE),
+    (0xf1, 0, columns::IS_CALL),
+    (0xf2, 0, columns::IS_CALLCODE),
+    (0xf3, 0, columns::IS_RETURN),
+    (0xf4, 0, columns::IS_DELEGATECALL),
+    (0xf5, 0, columns::IS_CREATE2),
+    (0xf6, 1, columns::IS_INVALID_16), // 0xf6-0xf7
+    (0xf8, 1, columns::IS_INVALID_17), // 0xf8-0xf9
+    (0xfa, 0, columns::IS_STATICCALL),
+    (0xfb, 0, columns::IS_INVALID_18),
+    (0xfc, 0, columns::IS_INVALID_19),
+    (0xfd, 0, columns::IS_REVERT),
+    (0xfe, 0, columns::IS_INVALID_20),
+    (0xff, 0, columns::IS_SELFDESTRUCT),
+];
+
+#[allow(dead_code)]
+pub fn generate<F: RichField>(lv: &mut [F; columns::NUM_CPU_COLUMNS]) {
+    let cycle_filter = lv[columns::IS_CPU_CYCLE];
+    if cycle_filter == F::ZERO {
+        return;
+    }
+    // This assert is not _strictly_ necessary, but I include it as a sanity check.
+    assert_eq!(cycle_filter, F::ONE, "cycle_filter should be 0 or 1");
+
+    let opcode = lv[columns::OPCODE].to_canonical_u64();
+    assert!(opcode < 256, "opcode should be in {{0, ..., 255}}");
+
+    for (i, &col) in columns::OPCODE_BITS.iter().enumerate() {
+        let bit = (opcode >> (7 - i)) & 1;
+        lv[col] = F::from_canonical_u64(bit);
+    }
+
+    let top_bits: [u64; 9] = [
+        0,
+        opcode & 0x80,
+        opcode & 0xc0,
+        opcode & 0xe0,
+        opcode & 0xf0,
+        opcode & 0xf8,
+        opcode & 0xfc,
+        opcode & 0xfe,
+        opcode,
+    ];
+
+    for (oc, block_length, col) in OPCODES {
+        lv[col] = F::from_bool(top_bits[8 - block_length] == oc);
+    }
+}
+
+pub fn eval_packed_generic<P: PackedField>(
+    lv: &[P; columns::NUM_CPU_COLUMNS],
+    yield_constr: &mut ConstraintConsumer<P>,
+) {
+    let cycle_filter = lv[columns::IS_CPU_CYCLE];
+
+    // Ensure that the opcode bits are valid: each has to be either 0 or 1, and they must match
+    // the opcode. Note that this also validates that this implicitly range-checks the opcode.
+    let bits = columns::OPCODE_BITS.map(|i| lv[i]);
+    // First check that the bits are either 0 or 1.
+    for bit in bits {
+        yield_constr.constraint(cycle_filter * bit * (bit - P::ONES));
+    }
+
+    // top_bits[i] is the opcode with all but the top i bits cleared.
+    let top_bits = {
+        let mut top_bits = [P::ZEROS; 9];
+        for i in 0..8 {
+            top_bits[i + 1] = top_bits[i] + bits[i] * P::Scalar::from_canonical_u64(1 << (7 - i));
+        }
+        top_bits
+    };
+
+    // Now check that they match the opcode.
+    let opcode = lv[columns::OPCODE];
+    yield_constr.constraint(cycle_filter * (opcode - top_bits[8]));
+
+    // Check that the instruction flags are valid.
+    // First, check that they are all either 0 or 1.
+    for &flag in lv
+        .iter()
+        .take(columns::END_INSTRUCTION_FLAGS)
+        .skip(columns::START_INSTRUCTION_FLAGS)
```

Comment:
A little more idiomatic IMO
```suggestion
    for &flag in &lv[columns::START_INSTRUCTION_FLAGS..columns::END_INSTRUCTION_FLAGS] {
```

Reply:
Oh true lol. (The chain of iterators was a mindless copy/paste from a Clippy suggestion)